### PR TITLE
Section Management

### PR DESCRIPTION
**Pull request by codecaffeine**
Added `appendSectionwithItems:` and `removeLastSection` to `URBNArrayDataSourceAdapter`
Added `appendSectionWithObject:items:` to `URBNAccordionDataSourceAdapter`.

If `isSection` isn’t set, these methods do nothing. I exposed section as a readonly Array (no mutable) because this information can be helpful.

### CODE REVIEW

File: Pod/Classes/URBNAccordionDataSourceAdapter.m [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
 @interface URBNAccordionDataSourceAdapter ()
 
-@property (nonatomic, strong) NSMutableArray *sections;
+@property (nonatomic, strong) NSArray *sections;
```
Comment (jgrandelli) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: Please redeclare with `readwrite`

File: Pod/Classes/URBNAccordionDataSourceAdapter.m [hunk 2]
```diff
@@ -28,7 +28,7 @@ - (instancetype)initWithSectionObjects:(NSArray *)sections andItems:(NSArray *)i
     if (self) {
         NSAssert(sections, @"You need sections for an accordion. Stop being a jerk.");
         NSAssert(sections.count > 0, @"Nice try, an empty sections array isn't gonna cut it. GTFO.");
-        self.sections = [NSMutableArray arrayWithArray:sections];
+        self.sections = [sections copy];
```
Comment (jgrandelli) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: Lets stick with array with array

File: Pod/Classes/URBNAccordionDataSourceAdapter.m [hunk 3]
```diff
@@ -117,6 +117,22 @@ - (BOOL)sectionIsOpen:(NSInteger)section {
     return [self.sectionsToKeepOpen containsIndex:section] || [self.expandedSections containsIndex:section];
 }
 
+- (void)appendSectionWithItems:(NSArray *)newItems {
+    NSAssert(YES, @"Call appendSectionObject:items:expanded: instead");
```
Comment (jgrandelli) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: Couldn't you just call through to `appendSectionObject:` and throw an assert in there for the section object? I think you should be asserting that you actually have a section in there anyways.

File: Pod/Classes/URBNArrayDataSourceAdapter.h [hunk 4]
```diff
@@ -34,6 +34,13 @@
 - (void)appendItems:(NSArray *)newItems inSection:(NSInteger)section;
 
 /**
+ * Appends a new section with the new items. If not sectioned, this does nothing.
```
Comment (jgrandelli) [inline Pod/Classes/URBNArrayDataSourceAdapter.h]: Does nothing? Why? Most of these methods manipulate the main items array if not sectioned. Would this not be desirable?

File: Pod/Classes/URBNArrayDataSourceAdapter.m [hunk 5]
```diff
@@ -93,6 +93,27 @@ - (void)appendItems:(NSArray *)newItems inSection:(NSInteger)section {
     }
 }
 
+- (void)appendSectionWithItems:(NSArray *)newItems {
+
```
Comment (jgrandelli) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: No \n here

File: Pod/Classes/URBNArrayDataSourceAdapter.m [hunk 6]
```diff
@@ -93,6 +93,27 @@ - (void)appendItems:(NSArray *)newItems inSection:(NSInteger)section {
...
+- (void)appendSectionWithItems:(NSArray *)newItems {
+    
+    if ([self isSectioned]) {
+        void (^updateData)() = ^{
+            [self.items addObject:newItems];
+        };
+        
+        if (self.tableView) {
+            updateData();
+            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:(self.items.count - 1)] withRowAnimation:UITableViewRowAnimationFade];
```
Comment (jgrandelli) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: Be sure to use `self.rowAnimation`

File: Pod/Classes/URBNArrayDataSourceAdapter.m [hunk 7]
```diff
@@ -93,6 +93,27 @@ - (void)appendItems:(NSArray *)newItems inSection:(NSInteger)section {
     }
 }
 
+- (void)appendSectionWithItems:(NSArray *)newItems {
+    
+    if ([self isSectioned]) {
```
Comment (jgrandelli) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: Also if you aren't going to perform an action lets throw an assert in there. So the consumer knows why it was a no-op and to stop doing it.

File: Pod/Classes/URBNArrayDataSourceAdapter.m [hunk 8]
```diff
@@ -260,6 +281,21 @@ - (void)removeItemAtIndexPath:(NSIndexPath *)indexPath {
     }
 }
 
+- (void)removeLastSection {
+    if ([self isSectioned] && (self.items.count > 0)) {
+        [self.items removeLastObject];
+
+        NSIndexSet *deletedSection = [NSIndexSet indexSetWithIndex:self.items.count];
+        if (self.tableView) {
+            [self.tableView deleteSections:deletedSection withRowAnimation:UITableViewRowAnimationFade];
```
Comment (jgrandelli) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: Be sure to use `self.rowAnimation`

File: Pod/Classes/URBNArrayDataSourceAdapter.m [hunk 9]
```diff
@@ -260,6 +281,21 @@ - (void)removeItemAtIndexPath:(NSIndexPath *)indexPath {
     }
 }
 
+- (void)removeLastSection {
+    if ([self isSectioned] && (self.items.count > 0)) {
```
Comment (jgrandelli) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: Same thing about the no-op assert

File: Pod/Classes/URBNArrayDataSourceAdapter.m [hunk 10]
```diff
@@ -260,6 +281,21 @@ - (void)removeItemAtIndexPath:(NSIndexPath *)indexPath {
...
+        [self.items removeLastObject];
+
+        NSIndexSet *deletedSection = [NSIndexSet indexSetWithIndex:self.items.count];
```
Comment (jgrandelli) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: Also i like how you put this outside the condition. Why not do the same with the append method

File: Pod/Classes/URBNAccordionDataSourceAdapter.m [hunk 11]
```diff
@@ -117,6 +117,22 @@ - (BOOL)sectionIsOpen:(NSInteger)section {
...
+}
+
+- (void)appendSectionObject:(id)sectionObject items:(NSArray *)items {
+    self.sections = [self.sections arrayByAddingObject:sectionObject];
```
Comment (jgrandelli) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: You should make sure that items is an actual array. You'll see in the initWithItems method we `arrayWithArray:items`. That way its A) not mutable and B) not inserting nil into the containing array

File: Pod/Classes/URBNArrayDataSourceAdapter.m (hunk 6 above)
Comment (codecaffeine) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: I actually tried `self.rowAnimation` and it caused the insertion to be pushed from the left which was ugly. In this case, would I change the datasource’s rowAnimation to be fade, perform the operation, then change it back?

File: Pod/Classes/URBNAccordionDataSourceAdapter.m (hunk 1 above)
Comment (codecaffeine) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: I’m just keeping this a private property and adding an `-allSections` to mimic `- allItems`

File: Pod/Classes/URBNAccordionDataSourceAdapter.m (hunk 2 above)
Comment (codecaffeine) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: done

File: Pod/Classes/URBNAccordionDataSourceAdapter.m (hunk 3 above)
Comment (codecaffeine) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: Done

File: Pod/Classes/URBNAccordionDataSourceAdapter.m (hunk 11 above)
Comment (codecaffeine) [inline Pod/Classes/URBNAccordionDataSourceAdapter.m]: done

File: Pod/Classes/URBNArrayDataSourceAdapter.h (hunk 4 above)
Comment (codecaffeine) [inline Pod/Classes/URBNArrayDataSourceAdapter.h]: done

File: Pod/Classes/URBNArrayDataSourceAdapter.m (hunk 9 above)
Comment (codecaffeine) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: done

File: Pod/Classes/URBNArrayDataSourceAdapter.m (hunk 10 above)
Comment (codecaffeine) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: done

File: Pod/Classes/URBNArrayDataSourceAdapter.m (hunk 8 above)
Comment (codecaffeine) [inline Pod/Classes/URBNArrayDataSourceAdapter.m]: done

**Comment by codecaffeine** [comment]
@jgrandelli updated from your comments

**Comment by jgrandelli** [comment]
Very nice I like it.

**Pull request merged**